### PR TITLE
PT-164657556 plant fortuna hard fork

### PR DESCRIPTION
**Pull request by hanssv**
This is preparing for the Fortuna hard fork. It introduces `ct-fortuna` and `eunit-fortuna` test targets. 

NOTE: *It does not have a good solution on how to turn Fortuna off for upcoming maintenance releases of Minerva* - suggestions are welcome, perhaps a little Makefile magic?! See comments below!

### CODE REVIEW

File: apps/aecontract/src/aect_contracts.erl [hunk 1]
```diff
@@ -129,18 +129,23 @@ is_legal_serialization_at_height(?ABI_SOPHIA_1, Vsn, Height) ->
...
-        ?MINERVA_PROTOCOL_VSN -> false
+        ?MINERVA_PROTOCOL_VSN -> false;
+        ?FORTUNA_PROTOCOL_VSN -> false
     end;
 is_legal_version_in_protocol(create, #{vm := ?VM_AEVM_SOPHIA_2, abi := ?ABI_SOPHIA_1}, ProtocolVersion) ->
     case ProtocolVersion of
         ?ROMA_PROTOCOL_VSN    -> false;
-        ?MINERVA_PROTOCOL_VSN -> true
+        ?MINERVA_PROTOCOL_VSN -> true;
+        ?FORTUNA_PROTOCOL_VSN -> true %% TODO: Revise this before release
```
Comment (hanssv) [inline apps/aecontract/src/aect_contracts.erl]: Mainly this is my worry!

File: apps/aecontract/src/aect_contracts.erl [hunk 2]
```diff
@@ -129,18 +129,23 @@ is_legal_serialization_at_height(?ABI_SOPHIA_1, Vsn, Height) ->
...
     end;
 is_legal_version_in_protocol(call, #{vm := VMVersion}, ProtocolVersion) ->
     case ProtocolVersion of
         ?ROMA_PROTOCOL_VSN    when VMVersion =:= ?VM_AEVM_SOPHIA_1 -> true;
         ?MINERVA_PROTOCOL_VSN when VMVersion =:= ?VM_AEVM_SOPHIA_1;
                                    VMVersion =:= ?VM_AEVM_SOPHIA_2 -> true;
+        %% TODO: Revise this before release
+        ?FORTUNA_PROTOCOL_VSN when VMVersion =:= ?VM_AEVM_SOPHIA_1;
+                                   VMVersion =:= ?VM_AEVM_SOPHIA_2 -> true;
```
Comment (hanssv) [inline apps/aecontract/src/aect_contracts.erl]: And this!

One could perhaps do some `ifdef` trickery?!

File: rebar.config [hunk 3]
```diff
@@ -119,6 +119,10 @@
                    {copy, "data/aecore/.minerva/accounts.json", "data/aecore/.minerva/accounts.json"},
                    {copy, "data/aecore/.minerva/accounts_uat.json", "data/aecore/.minerva/accounts_uat.json"},
                    {copy, "data/aecore/.minerva/accounts_test.json", "data/aecore/.minerva/accounts_test.json"},
+                   {mkdir, "data/aecore/.fortuna"},
```
Comment (lucafavatella) [inline rebar.config]: If this is here, then `ci/appveyor/package.bat` shall be amended too? cc @tolbrino

Comment (tolbrino) [reply rebar.config]: Yes.

File: apps/aecontract/src/aect_contracts.erl (hunk 1 above)
Comment (lucafavatella) [reply apps/aecontract/src/aect_contracts.erl]: The protocol version comes from network id or "chain > hard_forks" config, so this seems ok to me.

Could you please ensure that the comments marked TODO have a corresponding Pivotal Tracker ticket placed before Fortuna release placeholder?

File: apps/aecontract/src/aect_contracts.erl (hunk 2 above)
Comment (lucafavatella) [reply apps/aecontract/src/aect_contracts.erl]: The protocol version comes from network id or "chain > hard_forks" config, so this seems ok to me.

File: apps/aecore/src/aec_hard_forks.erl [hunk 4]
```diff
@@ -70,16 +71,25 @@ is_fork_height(Height) ->
 protocols_from_network_id(<<"ae_mainnet">>) ->
     #{ ?ROMA_PROTOCOL_VSN     => 0
      , ?MINERVA_PROTOCOL_VSN  => 47800
+     %% , ?FORTUNA_PROTOCOL_VSN => Not yet decided
```
Comment (lucafavatella) [inline apps/aecore/src/aec_hard_forks.erl]: 👍

File: apps/aecore/src/aec_hard_forks.erl [hunk 5]
```diff
@@ -70,16 +71,25 @@ is_fork_height(Height) ->
...
      };
 protocols_from_network_id(<<"ae_uat">>) ->
     #{ ?ROMA_PROTOCOL_VSN     => 0
      , ?MINERVA_PROTOCOL_VSN  => 40900
+     %% , ?FORTUNA_PROTOCOL_VSN => Not yet decided
```
Comment (lucafavatella) [inline apps/aecore/src/aec_hard_forks.erl]: 👍

File: apps/aecore/test/aec_mining_tests.erl [hunk 6]
```diff
@@ -114,7 +115,8 @@ generate_valid_test_data(TopBlock, Tries) ->
                             [{TopBlock, aec_trees:new()}], ?TEST_PUB),
     HeaderBin = aec_headers:serialize_to_binary(aec_blocks:to_header(BlockCandidate)),
     Target = aec_blocks:target(BlockCandidate),
-    case ?TEST_MODULE:generate(HeaderBin, Target, Nonce, 0) of
```
Comment (lucafavatella) [inline apps/aecore/test/aec_mining_tests.erl]: 🤔

File: Makefile [hunk 7]
```diff
@@ -195,6 +195,15 @@ ct-roma: internal-build
 		$(REBAR) ct $(CT_TEST_FLAGS) --sys_config config/test-roma.config; \
 	fi
 
+ct-fortuna: KIND=test
```
Comment (lucafavatella) [inline Makefile]: Circle CI config?

File: Makefile [hunk 8]
```diff
@@ -206,6 +215,10 @@ eunit-roma: KIND=test
 eunit-roma: internal-build
 	@ERL_FLAGS="-args_file $(EUNIT_VM_ARGS) -config $(EUNIT_SYS_CONFIG) -network_id local_roma_testnet" ./rebar3 do eunit $(EUNIT_TEST_FLAGS)
 
+eunit-fortuna: KIND=test
```
Comment (lucafavatella) [inline Makefile]: Circle CI config?

Comment (hanssv) [reply Makefile]: I wasn't sure we wanted to enable it immediately, the tests would be identical to the minerva tests until we actually start using it... But probably better to activate immediately.

File: apps/aecore/test/aec_mining_tests.erl (hunk 6 above)
Comment (hanssv) [reply apps/aecore/test/aec_mining_tests.erl]: aec_mining PR should have changed this, but this is code that is not run normally...

File: rebar.config (hunk 3 above)
Comment (hanssv) [reply rebar.config]: OK, will do... Didn't even know there was such a file... Why not use the marvelous rebar for this? 😂

File: apps/aecontract/src/aect_contracts.erl (hunk 1 above)
Comment (hanssv) [reply apps/aecontract/src/aect_contracts.erl]: https://www.pivotaltracker.com/story/show/164728984

**Comment by lucafavatella** [comment]
I am taking the liberty of labelling as WIP as not planned for this release - so to ease filtering of open PRs.

File: .circleci/config.yml [hunk 9]
```diff
@@ -878,6 +893,16 @@ workflows:
                 - env/dev2
                 - system-tests
 
+      - test_fortuna:
```
Comment (lucafavatella) [inline .circleci/config.yml]: We may consider to make this twice a day only cc @dincho

**Comment by lucafavatella** [comment]
@tolbrino Did you want to check anything else re Windows in this PR?

**Comment by lucafavatella** [comment]
I am removing WIP label following chat with @hanssv

File: rebar.config (hunk 3 above)
Comment (tolbrino) [reply rebar.config]: Don't make me do it O_O

**Pull request merged**